fix(joinCommunity): Fixing permissions checks in join community flows

alexjba opened the pull request:
### What does the PR do

Closing: #14095 
Closing: #14200
Closing: #12326


Moving the permissions model from communities module to chat section module (at least that's the main intention). But we'll need to also move or adapt other dependencies like check permissions on selected addresses, revealed addresses.

Motivation: The current implementation in communities module is built based on the assumption that we'll need the permissions for only one community at a time. But it's not the case. We have lots of qml bindings on the permission model, each one requesting the model for a new community because the permissionsModel access is done something like:

```
    readonly property var permissionsModel: {
        root.store.prepareTokenModelForCommunity(communityData.id)
        return root.store.permissionsModel
    }
```

O top of this, there are issues with permission update events handling and we're not always seeing the permission updates in qml.

### Affected areas

<!-- List the affected areas (e.g wallet, browser, etc..) -->

alexjba committed:
fix(joinCommunity): Fixing permissions checks in join community flows

alexjba commented:
@jrainville I'd really appreciate your thoughts here.
It's still very much WIP, but I need to confirm the direction with you because this change is snowballing and I don't want to waste time if I'm missing some crucial detail and we shouldn't move this implementation in the chat section. I can't really find the reason why it was implemented in the communities module and this is why I'm not 100% sure the direction is good.

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | d89a1930135dcacae668c5e08d7d3d00dd0a61f6 | [#1](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-14215/1/) | 2024-03-28 13:49:43 | ~6 min | `tests/nim` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-14215/1//consoleText) |

jrainville commented:
> @jrainville I'd really appreciate your thoughts here. It's still very much WIP, but I need to confirm the direction with you because this change is snowballing and I don't want to waste time if I'm missing some crucial detail and we shouldn't move this implementation in the chat section. I can't really find the reason why it was implemented in the communities module and this is why I'm not 100% sure the direction is good.

IIRC, the reason why we moved the models to the community module instead of the chat section one was because we use those models in the community portal as well. So even before the community is spectated, we want to be able to see what permissions they have.

![image](https://github.com/status-im/status-desktop/assets/11926403/5262ccb5-1803-41a1-83bb-6ecfb74fc278)

I think maybe because you could access the community from the settings screen too.

---

All in all, I agree that having the model is the chat section is way better, but it's also a lot of work and it will break some parts like community portal and probably something else I forgot.

IMO, this work should be done for 2.29 and be analysed well beforehand, because like you saw, it can snowball and then you realize too late that you've broken one of the modules by accident.

alexjba commented:
> > @jrainville I'd really appreciate your thoughts here. It's still very much WIP, but I need to confirm the direction with you because this change is snowballing and I don't want to waste time if I'm missing some crucial detail and we shouldn't move this implementation in the chat section. I can't really find the reason why it was implemented in the communities module and this is why I'm not 100% sure the direction is good.
> 
> IIRC, the reason why we moved the models to the community module instead of the chat section one was because we use those models in the community portal as well. So even before the community is spectated, we want to be able to see what permissions they have.
> 
> ![image](https://private-user-images.githubusercontent.com/11926403/317784680-5262ccb5-1803-41a1-83bb-6ecfb74fc278.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE2NDAyNzUsIm5iZiI6MTcxMTYzOTk3NSwicGF0aCI6Ii8xMTkyNjQwMy8zMTc3ODQ2ODAtNTI2MmNjYjUtMTgwMy00MWExLTgzYmItNmVjZmI3NGZjMjc4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI4VDE1MzI1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThmNGRiNDA5MWRmYWYzODBmNTY2MTllM2Y0M2JiMTJmMWQ4NDIyMzNhOTAzZjczYmYxMDM4NmUyNTc4MjQyZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TbuOFKxIORwuD036b5PzYgcxVo7loNFUGyduXgulZHs)
> 
> I think maybe because you could access the community from the settings screen too.
> 
> All in all, I agree that having the model is the chat section is way better, but it's also a lot of work and it will break some parts like community portal and probably something else I forgot.
> 
> IMO, this work should be done for 2.29 and be analysed well beforehand, because like you saw, it can snowball and then you realize too late that you've broken one of the modules by accident.

The models used in the community portal have the token permissions and also the tokens requirements met flag. So we might get away with it.
I've moved forward with this because the only place I've found it being used if for the join community flows. But I agree it should be done with less pressure in 2.29. The AFAIK the join community flows are already very fragile and this change has significant risks.

alexjba commented:
Closing this. It's been stalling for quite some time.
